src/slice.c

```diff
@@ -100,98 +103,136 @@ static SEXP vec_slice_fallback(SEXP x, SEXP index) {
                          syms_i, index);
 }
 
-static SEXP vec_slice_impl(SEXP x, SEXP index, bool dispatch) {
+static SEXP vec_slice_base(enum vctrs_type type,
+                           SEXP x,
+                           SEXP index,
+                           SEXP to, bool dispatch) {
+  SEXP out;
+  switch (type) {
+  case vctrs_type_logical:   out = lgl_slice(x, index); break;
+  case vctrs_type_integer:   out = int_slice(x, index); break;
+  case vctrs_type_double:    out = dbl_slice(x, index); break;
+  case vctrs_type_complex:   out = cpl_slice(x, index); break;
+  case vctrs_type_character: out = chr_slice(x, index); break;
+  case vctrs_type_raw:       out = raw_slice(x, index); break;
+  case vctrs_type_list:      out = list_slice(x, index); break;
+  default: Rf_error("Internal error: Non-base vector type `%s` in `vec_slice_base()`",
+                    vec_type_as_str(type));
+  }
+  PROTECT(out);
+
+  slice_names(out, x, index);
+  out = vec_restore(out, to, index);
+
+  UNPROTECT(1);
+  return out;
+}
+static void slice_names(SEXP x, SEXP to, SEXP index) {
+  SEXP nms = PROTECT(Rf_getAttrib(to, R_NamesSymbol));
+
+  if (nms == R_NilValue) {
+    UNPROTECT(1);
+    return;
+  }
+
+  nms = PROTECT(chr_slice(nms, index));
+  Rf_setAttrib(x, R_NamesSymbol, nms);
+
+  UNPROTECT(2);
+}
+
+/**
```

Comment:
Inconsistent commenting style

Reply:
I think we should use that style for docstrings with parameter lists, it's supported in vim and emacs.